### PR TITLE
Implement maxDepth in eager loading extension

### PR DESCRIPTION
**Pull request by hercemer42**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | na
| License       | MIT
| Doc PR        | na

### CODE REVIEW

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 1]
```diff
@@ -72,10 +79,11 @@ public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGenerator
 
         $forceEager = $this->shouldOperationForceEager($resourceClass, $options);
         $fetchPartial = $this->shouldOperationFetchPartial($resourceClass, $options);
+        $serializerContext = $this->getSerializerContext($resourceClass, static::NORMALIZATION_CONTEXT, $options);
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: `self::NORMALIZATION_CONTEXT` as this class is marked final? (Same everywhere).

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 2]
```diff
@@ -113,19 +121,23 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
+     * @param int                         $currentDepth            the current max depth
      *
      * @throws RuntimeException when the max number of joins has been reached
      */
-    private function joinRelations(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, bool $forceEager, bool $fetchPartial, string $parentAlias, array $propertyMetadataOptions = [], bool $wasLeftJoin = false, int &$joinCount = 0)
+    private function joinRelations(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, bool $forceEager, bool $fetchPartial, string $parentAlias, array $propertyMetadataOptions = [], array $context = [], bool $wasLeftJoin = false, int &$joinCount = 0, int $currentDepth = null)
     {
         if ($joinCount > $this->maxJoins) {
-            throw new RuntimeException('The total number of joined relations has exceeded the specified maximum. Raise the limit if necessary.');
+            throw new RuntimeException('The total number of joined relations has exceeded the specified maximum. Raise the limit if necessary, or implement @MaxDepth.');
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: `or use the "max_depth" option of the Symfony serializer`? (Max depth can be configured using a YAML or a XML config file too).

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 3]
```diff
@@ -154,6 +166,11 @@ private function joinRelations(QueryBuilder $queryBuilder, QueryNameGeneratorInt
                 $method = 'innerJoin';
             }
 
+            //Don't join if max depth is enabled and the current depth limit is reached
+            if (isset($context[static::ENABLE_MAX_DEPTH]) && $currentDepth === 0) {
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: `0 === $currentDepth` (we use Yoda style).

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 4]
```diff
@@ -174,7 +191,16 @@ private function joinRelations(QueryBuilder $queryBuilder, QueryNameGeneratorInt
                 continue;
             }
 
-            $this->joinRelations($queryBuilder, $queryNameGenerator, $mapping['targetEntity'], $forceEager, $fetchPartial, $associationAlias, $propertyMetadataOptions, $method === 'leftJoin', $joinCount);
+            if (null !== $attributesMetadata && isset($attributesMetadata[$association])) {
+                $maxDepth = $attributesMetadata[$association]->getMaxDepth();
+
+                // The current depth is the lowest max depth available in the ancestor tree.
+                if (null !== $maxDepth && ($maxDepth < $currentDepth || $currentDepth === null)) {
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: `null === $currentDepth`

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 5]
```diff
@@ -232,23 +258,32 @@ private function getSerializerGroups(string $resourceClass, array $options, stri
         }
 
         if (null !== $this->serializerContextBuilder && null !== $request) {
-            $contextFromRequest = $this->serializerContextBuilder->createFromRequest($request, $context === 'normalization_context');
-
-            if (isset($contextFromRequest['groups'])) {
-                return ['serializer_groups' => $contextFromRequest['groups']];
-            }
+            return $this->serializerContextBuilder->createFromRequest($request, $contextType === static::NORMALIZATION_CONTEXT);
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: `self::NORMALIZATION_CONTEXT === $contextType`

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 6]
```diff
@@ -38,20 +39,26 @@
 {
     use EagerLoadingTrait;
 
+    const ENABLE_MAX_DEPTH = 'enable_max_depth';
+    const NORMALIZATION_CONTEXT = 'normalization_context';
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: I don't think its a good idea to introduce does public constants here. They should at least be market with `@internal` (private constants are in PHP 7.1 only). But I'll remove them totally.

Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Ok, I've just removed the consts then

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 1 above)
Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: removed const

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 2 above)
Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Good point, changed.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 3 above)
Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: fixed

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 4 above)
Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: fixed

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 5 above)
Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: fixed

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 7]
```diff
@@ -113,19 +117,23 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
     {
         if ($joinCount > $this->maxJoins) {
-            throw new RuntimeException('The total number of joined relations has exceeded the specified maximum. Raise the limit if necessary.');
+            throw new RuntimeException('The total number of joined relations has exceeded the specified maximum. Raise the limit if necessary, or use the "max_depth" option of the Symfony serializer.');
         }
 
+        $currentDepth = $currentDepth > 0 ? $currentDepth - 1 : $currentDepth;
         $entityManager = $queryBuilder->getEntityManager();
         $classMetadata = $entityManager->getClassMetadata($resourceClass);
+        $attributesMetadata = null !== $this->classMetadataFactory ? $this->classMetadataFactory->getMetadataFor($resourceClass)->getAttributesMetadata() : null;
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]:
```php
$attributesMetadata = $this->classMetadataFactory ? $this->classMetadataFactory->getMetadataFor($resourceClass)->getAttributesMetadata() : null;
```

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 8]
```diff
@@ -154,6 +162,11 @@ private function joinRelations(QueryBuilder $queryBuilder, QueryNameGeneratorInt
                 $method = 'innerJoin';
             }
 
+            //Don't join if max depth is enabled and the current depth limit is reached
+            if (isset($context['enable_max_depth']) && 0 === $currentDepth) {
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Can you move this condition before the try/catch?

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 9]
```diff
@@ -174,7 +187,16 @@ private function joinRelations(QueryBuilder $queryBuilder, QueryNameGeneratorInt
                 continue;
             }
 
-            $this->joinRelations($queryBuilder, $queryNameGenerator, $mapping['targetEntity'], $forceEager, $fetchPartial, $associationAlias, $propertyMetadataOptions, $method === 'leftJoin', $joinCount);
+            if (null !== $attributesMetadata && isset($attributesMetadata[$association])) {
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]:
```php
if (isset($attributesMetadata[$association])) {
```

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 10]
```diff
@@ -174,7 +187,16 @@ private function joinRelations(QueryBuilder $queryBuilder, QueryNameGeneratorInt
...
+                $maxDepth = $attributesMetadata[$association]->getMaxDepth();
+
+                // The current depth is the lowest max depth available in the ancestor tree.
+                if (null !== $maxDepth && ($maxDepth < $currentDepth || null === $currentDepth)) {
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]:
```php
(null === $currentDepth || $maxDepth < $currentDepth)
```

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 8 above)
Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: yeah that works

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 10 above)
Comment (hercemer42) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: I tried your suggestion but it throws the max join error  as null < $currentDepth satisfies the condition, which we don't want.  I changed the order a bit to 
(null !== $maxDepth && (null === $currentDepth || $maxDepth < $currentDepth))

Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Yeah that's what I meant: `null !== $maxDepth && (null === $currentDepth || $maxDepth < $currentDepth))` 😉

**Comment by dunglas** [comment]
On hold before the release of 2.1.

**Comment by dunglas** [comment]
As they are already many change in the beta 2, let's include this one.

**Comment by dunglas** [comment]
Thanks @hercemer42!

**Pull request merged**